### ISSUE TITLE
Issue with crypto/ocb.cc on Mac OS X 10.8 related to CommonCryptor.h

### ISSUE THREAD

**Issue by Marcool04** [mobile-shell/mosh#804]
Hi,
I had an issue today while installing mosh on a (rather vintage I'll admit...) Macbook 4,1 running OS X 10.8. I tried 
```
$ brew install mobile-shell
[TRUNCATED]
==> ./configure --prefix=/usr/local/Cellar/mobile-shell/1.2.6 --enable-completion
==> make check
Last 15 lines from /Users/marcool/Library/Logs/Homebrew/mobile-shell/02.make:
    kCCAlgorithmDES,        
    ^
ocb.cc:409:3: error: use of undeclared identifier 'kCCAlgorithmAES'; did you mean 'kCCAlgorithmDES'?
                kCCAlgorithmAES,
                ^~~~~~~~~~~~~~~
                kCCAlgorithmDES
/usr/include/CommonCrypto/CommonCryptor.h:151:5: note: 'kCCAlgorithmDES' declared here
    kCCAlgorithmDES,        
    ^
2 errors generated.
make[3]: *** [ocb.o] Error 1
make[3]: *** Waiting for unfinished jobs....
make[2]: *** [check-recursive] Error 1
make[1]: *** [check-recursive] Error 1
make: *** [check] Error 2
```

I then downloaded the 1.2.6 release from git, and did this:
 
```
[CommonCrypto.zip](https://github.com/mobile-shell/mosh/files/461099/CommonCrypto.zip)
$ ./autogen.sh 
$ ./configure
[TRUNCATED]
checking CommonCrypto/CommonCrypto.h usability... yes
checking CommonCrypto/CommonCrypto.h presence... yes
checking for CommonCrypto/CommonCrypto.h... yes
checking for CommonCrypto/CommonCrypto.h... (cached) yes
[TRUNCATED]
configure:  === Configuration results ===
configure: Version:             1.2.6
configure: c++ compiler:        c++
configure: Warning CXXFLAGS:    -Wall
configure: Picky CXXFLAGS:      
configure: Harden CFLAGS:        -fno-strict-overflow -D_FORTIFY_SOURCE=2 -fstack-protector-all -Wstack-protector --param ssp-buffer-size=1 -fPIE
configure:  =============================
$ make
make  all-recursive
Making all in scripts
Making all in src
Making all in protobufs
make  all-am
make[4]: Nothing to be done for `all-am'.
Making all in util
make[3]: Nothing to be done for `all'.
Making all in crypto
  CXX      ocb.o
ocb.cc:393:3: error: use of undeclared identifier 'kCCAlgorithmAES'; did you mean 'kCCAlgorithmDES'?
                kCCAlgorithmAES,
                ^~~~~~~~~~~~~~~
                kCCAlgorithmDES
/usr/include/CommonCrypto/CommonCryptor.h:151:5: note: 'kCCAlgorithmDES' declared here
    kCCAlgorithmDES,        
    ^
ocb.cc:409:3: error: use of undeclared identifier 'kCCAlgorithmAES'; did you mean 'kCCAlgorithmDES'?
                kCCAlgorithmAES,
                ^~~~~~~~~~~~~~~
                kCCAlgorithmDES
/usr/include/CommonCrypto/CommonCryptor.h:151:5: note: 'kCCAlgorithmDES' declared here
    kCCAlgorithmDES,        
    ^
2 errors generated.
make[3]: *** [ocb.o] Error 1
make[2]: *** [all-recursive] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all] Error 2
```

After looking in /usr/includes/CommonCrypto/CommonCryptor.h I realised that there was no mention of the constant kCCAlgorithmAES, but instead there was kCCAlgorithmAES128. I added 128 onto the end of all references to that constant in ocb.cc in mosh source, and compilation was successful.
I'm not sure if this is due to my system having a "weird" or otherwise non standard version of the CommonCrypto library? I'll include a copy of my CommonCrypto folder just in case.

Thanks for all the work on this great shell by the way!
Best regards,
Mark.

[CommonCrypto.zip](https://github.com/mobile-shell/mosh/files/461101/CommonCrypto.zip)

**Comment by cgull** [comment]
The relevant diff between that and the Xcode 7.3.1/SDK 10.11 install I have in front of me is:

```diff
@@ -138,8 +114,12 @@
 /*!
     @enum       CCAlgorithm
     @abstract   Encryption algorithms implemented by this module.
-
+
     @constant   kCCAlgorithmAES128  Advanced Encryption Standard, 128-bit block
+                                    This is kept for historical reasons.  It's
+                                    preferred now to use kCCAlgorithmAES since
+                                    128-bit blocks are part of the standard.
+    @constant   kCCAlgorithmAES     Advanced Encryption Standard, 128-bit block
     @constant   kCCAlgorithmDES     Data Encryption Standard
     @constant   kCCAlgorithm3DES    Triple-DES, three key, EDE configuration
     @constant   kCCAlgorithmCAST    CAST
@@ -148,7 +128,8 @@
 */
 enum {
     kCCAlgorithmAES128 = 0,
-    kCCAlgorithmDES,
+    kCCAlgorithmAES = 0,
+    kCCAlgorithmDES,
     kCCAlgorithm3DES,
     kCCAlgorithmCAST,
     kCCAlgorithmRC4,
```

So there is an incompatibility there and your workaround should be OK.

I built the Mosh 1.2.6 package for OS X 9 (though I'm not 100% sure it used the 10.9 SDK), and it appears that you can install Xcode 5.1.1 which also has the OS X 9 SDK.  You might have better luck with that.

I'm not sure yet what the best thing for us to do is yet.

**Comment by Marcool04** [comment]
Hi! Thanks for looking into this so quickly. I'm afraid that I already have XCode 5.1.1:
```
Developer Information:

  Version:	5.1.1 (5B1008)
  Location:	/Applications/Utilities/Xcode.app
  Applications:
  Xcode:	5.1.1 (5085)
  Instruments:	5.1.1 (55045)
  SDKs:
  OS X:
  10,8:	(12F37)
  10,9:	(13C64)
  iOS:
  7,1:	(11D167)
  iOS Simulator:
  7,1:	(11D167)
```
So I guess the fix (if you want to ensure such backward backward compatibility with 10.8...) would have to be a check for which CommonCrypto is being used :/ Or even a dirty `if (not defined kCCAlgorithmAES) use kCCAlgorithmAES128` some place in ocb.cc

**Comment by awesomer** [comment]
I was able to work around this issue and get mosh working on homebrew on mountain lion by doing the following :

1) brew unpack mosh
2) cd ~/Library/Caches/Homebrew/mobile-shell-1.2.6
3) cp ./src/crypto/ocb.cc ./src/crypto/ocb.cc.fixed
4) vi ./src/crypto/ocb.cc.fixed
5) replace two instances of kCCAlgorithmAES with kCCAlgorithmAES128
6) diff -u ./src/crypto/ocb.cc ./src/crypto/ocb.cc.fixed | pbcopy
7) pasted into a gist with the "raw" url at https://gist.githubusercontent.com/anonymous/8e93e32f6d58cfafd51b8ab51eaafdb2/raw/ff47f1713ee6f23533dc629dca6683dfd53615fc/gistfile1.txt
8) brew edit mosh
right above "def install", I added the following
"
  def patches
       "https://gist.githubusercontent.com/anonymous/8e93e32f6d58cfafd51b8ab51eaafdb2/raw/ff47f1713ee6f23533dc629dca6683dfd53615fc/gistfile1.txt"
  end
"
9) brew install mosh
10) modify /usr/local/bin/mosh to use /usr/local/bin/perl instead of env perl

It's possible that step 10 is something else that's broken about stock perl paths on homebrew in 10.8.5, but the above got me to a working state. Commenting here in the hopes it might help someone with the same issue.

**Comment by ilovezfs** [comment]
This will also do the trick
```
diff --git a/Formula/mobile-shell.rb b/Formula/mobile-shell.rb
index 58b934f..793f151 100644
--- a/Formula/mobile-shell.rb
+++ b/Formula/mobile-shell.rb
@@ -31,6 +31,10 @@ class MobileShell < Formula
     # PATH to support launching outside shell e.g. via launcher
     inreplace "scripts/mosh.pl", "'mosh-client", "\'#{bin}/mosh-client"
 
+    if MacOS.version <= :mountain_lion
+      inreplace "src/crypto/ocb.cc", "kCCAlgorithmAES", "kCCAlgorithmAES128"
+    end
+
     # Upstream prefers O2:
     # https://github.com/keithw/mosh/blob/master/README.md
     ENV.O2

```
but an upstream fix would be much appreciated!

**Comment by Marcool04** [comment]
Thanks for the clean patch data. Both of those worked for me too.
Guess we'll leave it up to the mosh devs to see if they want to tweak mosh to get it to work this upstream or not.
All the best.
Mark.

**Comment by cgull** [comment]
This is fixed in f429fd38b07a6d3bdd2e771924276e1f31a6c3b6 on master.  Using the old name is the right thing to do here.  There's a comment in `CommonCryptor.h` recommending the new name, and I originally followed that recommendation-- but it's unlikely Apple will ever actually deprecate the old name.

@ilovezfs:  Hmm, that patch to the Homebrew formula will now do a bad substitution here.  Did that ever make it into the Homebrew retro-OS forks?

**Comment by ilovezfs** [comment]
@cgull nope, that diff was only something I posted here, not something I ever added to the formula.
